lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/Restoration.hs

```diff
@@ -0,0 +1,134 @@
+import Data.Quantity
+    ( Quantity (..)
+    )
+import Test.Hspec
+    ( SpecWith
+    , describe
+    )
+import Test.Integration.Framework.DSL
+    ( Context
+    , eventually
+    )
+import Test.Integration.Framework.DSL.TestM
+    ( itM
+    , over
+    , partialRequest
+    )
+import Test.Integration.Framework.DSL.Wallet
+    ( Patch
+    , balanceIs
+    , createARandomWallet
+    , createARandomWalletWithMnemonics
+    , createAWallet
+    , deleteWallet
+    , fundWallet
+    , named
+    , waitUntilStateIsReady
+    , withApiWallet
+    )
+
+import qualified Cardano.Wallet.Api.Clients.Testnet.Network as C
+
+spec :: SpecWith Context
+spec = describe "restoration of wallets" $ do
+    itM "WALLET_RESTORE_0.1 create a wallet restoring from tip" $ do
+        w <- createARandomWallet $ named "Wallet from tip" . restoringFromTip
+        over w $ do
+            waitUntilStateIsReady
+            withApiWallet $ balanceIs 0
+
+    itM "WALLET_RESTORE_0.2 create a wallet restoring from genesis" $ do
+        w <-
+            createARandomWallet
+                $ named "Wallet from genesis" . restoringFromGenesis
+        over w $ do
+            waitUntilStateIsReady
+            withApiWallet $ balanceIs 0
+
+    itM "WALLET_RESTORE_0.3 create a wallet restoring from a checkpoint" $ do
+        cp <- partialRequest C.blocksLatestHeader
+        w <-
+            eventually "Wallet from a block"
+                $ createARandomWallet
+                $ named "Wallet from a block" . restoringFromCheckpoint cp
+        over w $ do
+            waitUntilStateIsReady
+            withApiWallet $ balanceIs 0
+
+    itM "WALLET_RESTORE_0.4 create a wallet from the tip \
+            \ignores past transactions" $ do
+        (genesisClone, mnemonics) <-
+            createARandomWalletWithMnemonics
+                $ named "Wallet from genesis" . restoringFromGenesis
+        over genesisClone $ do
+            waitUntilStateIsReady
+            fundWallet 42_000_000
+            withApiWallet $ balanceIs 42_000_000
+            deleteWallet
+        tipClone <-
+            createAWallet mnemonics
+                $ named "Wallet from tip" . restoringFromTip
+        over tipClone $ do
+            waitUntilStateIsReady
+            withApiWallet $ balanceIs 0
+
+    itM "WALLET_RESTORE_0.5 create a wallet from the genesis \
+            \capture the past transactions" $ do
+        (genesisClone, mnemonics) <-
+            createARandomWalletWithMnemonics
+                $ named "Wallet from genesis" . restoringFromTip
+        over genesisClone $ do
+            waitUntilStateIsReady
+            fundWallet 42_000_000
+            withApiWallet $ balanceIs 42_000_000
+            deleteWallet
+        tipClone <-
+            createAWallet mnemonics
+                $ named "Wallet from genesis second take" . restoringFromGenesis
+        over tipClone $ do
+            waitUntilStateIsReady
+            withApiWallet $ balanceIs 42_000_000
```

Comment:
what about adding the following integration tests:
1.  (a) creating wallet and funding it. (b) taking slotNO/headerhash T (c) waiting 2 epochs. 
-> show  that I can restore from T and got proper balance
-> show that I can restore from genesis and got proper balance
-> show that restoring from tip gives 0 balance
2. (a) create other tx, for example join pool.
-> show like in 1 that when restoring I can get certificates/delegating correctly depending on restoration mode

Reply:
3. Demonstrate ErrCreateWalletRestorationFromABlockFailed